src/airflow/airqo_etl_utils/airqo_utils.py

```diff
@@ -964,9 +967,28 @@ def flag_faults(df):
 
     @staticmethod
     def save_faulty_devices(data: pd.DataFrame):
-        """Save faulty devices to MongoDB"""
-        client = pm.MongoClient(configuration.MONGO_URI)
-        db = client[configuration.MONGO_DATABASE_NAME]
-        records = data.to_dict("records")
-        db.faulty_devices.insert_many(records)
-        print("Faulty devices saved to MongoDB")
+        """Save or update faulty devices to MongoDB"""
+        with pm.MongoClient(configuration.MONGO_URI) as client:
+            db = client[configuration.MONGO_DATABASE_NAME]
+            records = data.to_dict("records")
+
+            bulk_ops = [
+                pm.UpdateOne(
+                    {"device_name": record["device_name"]},
+                    {"$set": record},
+                    upsert=True,
+                )
+                for record in records
+            ]
+
+            try:
+                db.faulty_devices.bulk_write(bulk_ops)
+            except Exception as e:
+                print(f"Error saving faulty devices to MongoDB: {e}")
+
+            current_device_names = [record["device_name"] for record in records]
+            db.faulty_devices.delete_many(
```

Comment:
Hi @Mnoble-19 , please shed more light on this delete many operation?
If possible, please use PR review comments on all the changed files to make our review process easier?
Thanks!